src/routes/getSkipSegments.js

```diff
@@ -8,271 +8,179 @@ var privateDB = databases.privateDB;
-
-  let finalSponsorTimeIndexes = weightedRandomIndexes.finalChoices;
-  //the sponsor times either chosen to be added to finalSponsorTimeIndexes or chosen not to be added
-  let finalSponsorTimeIndexesDealtWith = weightedRandomIndexes.choicesDealtWith;
-
-  let voteSums = weightedRandomIndexes.weightSums;
-  //convert these into the votes
-  for (let i = 0; i < finalSponsorTimeIndexes.length; i++) {
-      //it should use the sum of votes, since anyone upvoting a similar sponsor is upvoting the existence of that sponsor.
-      votes[finalSponsorTimeIndexes[i]] = voteSums[i];
-  }
-
-  //find the indexes never dealt with and add them
-  for (let i = 0; i < sponsorTimes.length; i++) {
-      if (!finalSponsorTimeIndexesDealtWith.includes(i)) {
-          finalSponsorTimeIndexes.push(i)
-      }
-  }
-
-  //if there are too many indexes, find the best 4
-  if (finalSponsorTimeIndexes.length > 8) {
-      finalSponsorTimeIndexes = getWeightedRandomChoice(finalSponsorTimeIndexes, votes, 8).finalChoices;
-  }
+//Segments with less than -1 votes are already ignored before this function is called
+function chooseSegments(segments) {
+  //Create groups of segments that are similar to eachother
+  //Segments must be sorted by their startTime so that we can build groups chronologically:
+  //1. As long as the segments' startTime fall inside the currentGroup, we keep adding them to that group
+  //2. If a segment starts after the end of the currentGroup (> cursor), no other segment will ever fall
+  //   inside that group (because they're sorted) so we can create a new one
+  const similarSegmentsGroups = [];
+  let currentGroup;
+  let cursor = -1; //-1 to make sure that, even if the 1st segment starts at 0, a new group is created
+  segments.forEach(segment => {
+    if (segment.startTime > cursor) {
+      currentGroup = { segments: [], votes: 0 };
+      similarSegmentsGroups.push(currentGroup);
+    }
 
-  //convert this to a final array to return
-  let finalSponsorTimes = [];
-  for (let i = 0; i < finalSponsorTimeIndexes.length; i++) {
-      finalSponsorTimes.push(sponsorTimes[finalSponsorTimeIndexes[i]]);
-  }
+    currentGroup.segments.push(segment);
+    //only if it is a positive vote, otherwise it is probably just a sponsor time with slightly wrong time
+    if (segment.votes > 0) {
+      currentGroup.votes += segment.votes;
+    }
 
-  //convert this to a final array of UUIDs as well
-  let finalUUIDs = [];
-  for (let i = 0; i < finalSponsorTimeIndexes.length; i++) {
-      finalUUIDs.push(UUIDs[finalSponsorTimeIndexes[i]]);
-  }
+    cursor = Math.max(cursor, segment.endTime);
+  });
 
-  return {
-      sponsorTimes: finalSponsorTimes,
-      UUIDs: finalUUIDs
-  };
+  //if there are too many groups, find the best 8
+  return getWeightedRandomChoice(similarSegmentsGroups, 8).map(
+    //randomly choose 1 good segment per group and return them
+    group => getWeightedRandomChoice(group.segments, 1)[0]
+  );
```

Comment:
I forgot about this change. When submitting, I removed this limit. I feel it's probably a bad idea to not have any limit here, but maybe it should just be raised a bit? What do you think

Reply:
Do you know if there are a lot of videos with more than 8 (separate) sponsors? I guess if I saw more than 8 sponsors as a viewer, I'd probably think that there's something wrong anyway.

Apart from that, I'd suggest to make any changes that actually change the behaviour in another PR/commit so that if there's something wrong with this PR, you can safely revert it without worrying about behaviour changes, if that makes sense.